File: Bus-Gogogo/MainViewController.m
//
//  MainControllerViewController.m
//  Bus-Gogogo
//
//  Created by Weijing Liu on 9/27/14.
//  Copyright (c) 2014 Weijing Liu. All rights reserved.
//

#import "MainViewController.h"
#import "PTMainTableViewCell.h"
#import "PTBase.h"
#import "PTDownloadTask.h"
#import "PTStopsForLocationRequester.h"
#import "PTStopMonitoringDownloadRequester.h"

static const NSString *kMainTableViewCellIdentifier = @"MainTableViewCell";
typedef enum : NSUInteger {
  MainTableViewCellTagLeftHead = 10,
  MainTableViewCellTagTitle,
  MainTableViewCellTagSubtitle,
} MainTableViewCellTag;

@interface MainViewController () <
  CLLocationManagerDelegate,
  UITableViewDataSource,
  UITableViewDelegate
>
{
  __weak IBOutlet UITableView *_tableView;
  
  CLLocationManager *_locationManager;
  CLLocation *_location;
  
  NSArray *_routeStopPairs;
}

@end

@implementation MainViewController

- (void)awakeFromNib
{
  [super awakeFromNib];
  
  _locationManager = [CLLocationManager new];
  _locationManager.delegate = self;
  
  [self _startPollLocation];
}

- (void)viewDidLoad
{
  [super viewDidLoad];
  
  _tableView.tableFooterView = [UIView new];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
  return _routeStopPairs.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
  PTMainTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([PTMainTableViewCell class])];
  RouteStopPair *pair = [_routeStopPairs objectAtIndex:indexPath.row];
  cell.head.text = pair.route.shortName;
  cell.head.textColor = pair.route.textColor;
  cell.head.backgroundColor = pair.route.color;
  //TODO:Still couldn't get the destination
  cell.title.text = [@"To " stringByAppendingString:[pair.route.destinations objectAtIndex:pair.stop.directionText.length % 2]];
  
  NSArray *component = @[[PTBase distanceStringBetweenA:_location b:pair.stop.location],
                         pair.stop.directionText,
                         pair.stop.name];
  cell.subtitle.text = [component componentsJoinedByString:@", "];
  return cell;
}

#pragma mark - private

- (void)_fetchStopsForLocation:(CLLocation *)location
{
  [PTDownloadTask
   scheduledTaskWithRequester:[[PTStopsForLocationRequester alloc] initWithLocation:location]
   callback:^(PTStopsForLocationRequester *requester, NSError *error) {
     assert(!error);
     _routeStopPairs = [self _sortRouteStopPairsByDistance:requester.routeStopPairs];
     [_tableView reloadData];
   }];
}

- (void)_fetchStopMonitoringData:(NSArray *)stopIds
{
  for (NSString *identifier in stopIds) {
    [PTDownloadTask scheduledTaskWithRequester:
     [[PTStopMonitoringDownloadRequester alloc] initWithStopId:identifier routeId:nil]
                                      callback:
     ^(PTStopMonitoringDownloadRequester *requester, NSError *error) {
       assert(!error);
       [requester monitoredJourneys];
     }];
  }
}

- (NSArray *)_sortRouteStopPairsByDistance:(NSArray *)routeStopPairs
{
  // routeId, direction,
  routeStopPairs = [routeStopPairs sortedArrayUsingComparator:^NSComparisonResult(RouteStopPair *a, RouteStopPair *b) {
    if ([a.route.identifier isEqualToString:b.route.identifier]) {
      if ([a.stop.directionText isEqualToString:b.stop.directionText]) {
        return fabs([a.stop.location distanceFromLocation:_location]) >
          fabs([b.stop.location distanceFromLocation:_location]);
      }
      return [a.stop.directionText compare:b.stop.directionText];
    }
    return [a.route.identifier compare:b.route.identifier];
  }];
  
  NSMutableArray *tmp = [NSMutableArray new];
  for (RouteStopPair *p in routeStopPairs) {
    RouteStopPair *q = tmp.lastObject;
    if (![p.route.identifier isEqualToString:q.route.identifier] &&
        ![p.stop.directionText isEqualToString:q.stop.directionText]) {
      [tmp addObject:p];
    }
  }
  return tmp;
}

- (void)_startPollLocation
{
  _locationManager.desiredAccuracy = kCLLocationAccuracyBest;
  [_locationManager startUpdatingLocation];
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray *)locations
{
  _location = locations.firstObject;
  [manager stopUpdatingLocation];
  if (locations.firstObject) {
    [self _fetchStopsForLocation:locations.firstObject];
  }
}

@end

File: Bus-Gogogo/MainViewController.swift
//
//  MainViewController.swift
//  Bus-Gogogo
//
//  Created by Weijing Liu on 9/29/14.
//  Copyright (c) 2014 Weijing Liu. All rights reserved.
//

import Foundation
import UIKit
import CoreLocation

class MainViewController:
  UIViewController,
  CLLocationManagerDelegate,
  UITableViewDataSource,
  UITableViewDelegate {

  @IBOutlet weak var tableView: UITableView! {
    didSet {
      tableView.tableFooterView = UIView()
    }
  }
  
  var _locationManager: CLLocationManager = CLLocationManager()
  
  var _location: CLLocation?
  
  var _closestRoutes: [ClosestRoute] = [ClosestRoute]()
  
  var _closestRouteMonitor: [RouteStopKey: String] = [RouteStopKey: String]()
  
  var _timer: NSTimer?
  
  class func key(r: PTRoute, s: PTStop) -> String {
    return r.identifier() + s.identifier
  }
  
  required init(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    _locationManager.delegate = self;
  }
  
  private func fetchClosestRoute() {
    ClosestRouteProvider(location: _location) { (closestRoute, error) -> () in
      assert(error == nil, "should be nil")
      self._closestRoutes = closestRoute!
      self.tableView.reloadData()
      self.fetchMonitor()
    }
  }
  
  @objc private func fetchMonitor() {
    for c in _closestRoutes {
      RestTask(requester: RestStopMonitoringRequester(monitoringRef: c.stop.identifier, lineRef: c.route.identifier())).start
      { (req: RestStopMonitoringRequester, error: NSError!) -> () in
        var key = RouteStopKey(r: c.route, s: c.stop)
        var journey = req._monitoredJourneys.first?
        
        self._closestRouteMonitor[key] = journey?.presentableDistance
        self.tableView.reloadData()
      }
    }
  }
  
  //MARK: - Private
  private func getClosestRoute(indexPath: NSIndexPath) -> ClosestRoute {
    return _closestRoutes[indexPath.row]
  }
  
  //MARK: - UIViewController
  override func viewWillAppear(animated: Bool) {
    super.viewWillAppear(animated)
    
    _locationManager.startUpdatingLocation()
    
    _timer?.invalidate()
    _timer = NSTimer.scheduledTimerWithTimeInterval(20, target: self, selector: Selector("fetchMonitor"), userInfo: nil, repeats: true)
  }
  
  override func viewWillDisappear(animated: Bool) {
    super.viewWillDisappear(animated)
    _timer?.invalidate()
  }
  
  //MARK: - CLLocationManagerDelegate
  func locationManager(manager: CLLocationManager!, didUpdateLocations locations: [AnyObject]!) {
    _locationManager.stopUpdatingLocation()
    _location = locations.first as? CLLocation
    fetchClosestRoute()
  }
  
  //MARK: - UITableViewDataSource
  func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return _closestRoutes.count
  }
  
  func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
    let closestRoute = _closestRoutes[indexPath.row]
    
    var identifier = NSStringFromClass(MainTableViewCell.classForCoder())
    var cell = tableView.dequeueReusableCellWithIdentifier(
      identifier, forIndexPath: indexPath) as MainTableViewCell
    cell.head.text = _closestRoutes[indexPath.row].route.shortName()
    cell.head.textColor = _closestRoutes[indexPath.row].route.textColor()
    cell.head.backgroundColor = _closestRoutes[indexPath.row].route.color()
    var dest = _closestRoutes[indexPath.row].destination
    let location = dest.rangeOfString("via").location
    if location != NSNotFound {
      dest = dest.substringToIndex(location)
    }
    cell.title.text = "To " + dest
    
    let monitor = _closestRouteMonitor[RouteStopKey(r: closestRoute.route, s: closestRoute.stop)]
    cell.subtitle.text = _closestRoutes[indexPath.row].stop.name
    cell.subline.text = (monitor == nil) ? "Not Available" : monitor!
    
    return cell
  }
  
  func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
    let closestRoute = getClosestRoute(indexPath)
    let cell = tableView.cellForRowAtIndexPath(indexPath) as MainTableViewCell
    
    cell.progressBar.onTap()
    if (cell.progressBar.state == CircularProgressBar.State.Loading) {
      PTReminderManager(stop: closestRoute.stop, route: closestRoute.route, stopsAway:3)
    }
  }
  
}

class RouteStopKey: Hashable, Equatable {
  var key: String
  required init(r: PTRoute, s: PTStop) {
    key = r.identifier() + s.identifier
  }
  
  var hashValue: Int {
    return key.hashValue
  }
}
func ==(lhs: RouteStopKey, rhs: RouteStopKey) -> Bool {
  return lhs.hashValue == rhs.hashValue
}
    